contracts/src/v0.8/dev/CanaryUpkeep.sol

```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.13;
+
+import "../interfaces/KeeperCompatibleInterface.sol";
+import "../interfaces/KeeperRegistryInterface.sol";
+import "../ConfirmedOwner.sol";
+
+/**
+ * @notice A canary upkeep which requires a different keeper to service its upkeep at an interval. This makes sure that
+ * all keepers are in a healthy state.
+ */
+contract CanaryUpkeep is KeeperCompatibleInterface, ConfirmedOwner {
+  uint256 private s_keeperIndex;
+  uint256 private s_interval = 300;
+  uint256 private s_timestamp;
+  KeeperRegistryInterface private immutable s_keeperRegistry;
+
+  /**
+   * @param keeperRegistry address of a keeper registry
+   */
+  constructor(KeeperRegistryInterface keeperRegistry) ConfirmedOwner(msg.sender) {
+    s_keeperRegistry = keeperRegistry;
+    s_timestamp = block.timestamp;
+    s_keeperIndex = 0;
+  }
+
+  /**
+   * @return the current keeper index
+   */
+  function getKeeperIndex() external view returns (uint256) {
+    return s_keeperIndex;
+  }
+
+  /**
+   * @return the current timestamp
+   */
+  function getTimestamp() external view returns (uint256) {
+    return s_timestamp;
+  }
+
+  /**
+   * @return the current interval
+   */
+  function getInterval() external view returns (uint256) {
+    return s_interval;
+  }
+
+  /**
+   * @notice updates the interval
+   * @param interval the new interval
+   */
+  function setInterval(uint256 interval) external onlyOwner {
+    s_interval = interval;
+  }
+
+  /**
+   * @notice returns true if keeper array is not empty and sufficient time has passed
+   */
+  function checkUpkeep(
+    bytes calldata /* checkData */
+  ) external view returns (bool, bytes memory) {
+    (State memory _s, Config memory _c, address[] memory keepers) = s_keeperRegistry.getState();
+    bool upkeepNeeded = keepers.length != 0 && block.timestamp >= s_interval + s_timestamp;
+    return (upkeepNeeded, bytes(""));
+  }
+
+  /**
+   * @notice checks keepers array limit, timestamp limit, and requires transaction origin must be the anticipated keeper.
+   * If all checks pass, update the keeper index and timestamp. Otherwise, revert this transaction.
+   */
+  function performUpkeep(
+    bytes calldata /* performData */
+  ) external {
+    (State memory _s, Config memory _c, address[] memory keepers) = s_keeperRegistry.getState();
+    if (keepers.length == 0) {
+      revert("no keeper nodes exists");
+    }
+    if (block.timestamp < s_interval + s_timestamp) {
+      revert("Not enough time has passed after the previous upkeep");
+    }
+    // if keepers array is shortened, this statement will make sure keeper index is always valid
+    if (s_keeperIndex >= keepers.length) {
+      s_keeperIndex = 0;
+    }
+
+    require(tx.origin == keepers[s_keeperIndex], "transaction origin is not the anticipated keeper.");
+    s_keeperIndex = (s_keeperIndex + 1) % keepers.length;
```

Comment:
not sure if this will really work. Keepers will rotate turns during BCPT, so we can't assume the next keeper would be the next one.

How are we planning on monitoring this exactly?

Reply:
hmm, i don't think we can predict which the next keeper will be, other than the fact that it will be a different keeper.

Though we can probably say that over a long period of time, all keepers should perform this equal number of times

Reply:
this is exactly what we just spent time undoing. We wanted it to be not in a predictable circle. I think we need another way to think about this.

Reply:
from what I understand this will show three people as keepers of the registry but will also keep the owner of the contracts (me) from being able to access any of the information within, am I correct in my assessment?

Reply:
@infiloop2 @cmalec I think this is okay as long as the alert interval is long enough such that probabilistically, every keeper should have seen the canary at least once during this period.